src/org/python/core/JyAttribute.java

```diff
@@ -197,18 +193,51 @@ public static boolean hasAttr(PyObject ob, byte attr_type) {
      * Retrieves the attribute of the given type from the given
      * {@link org.python.core.PyObject}.
      * If no attribute of the given type is attached, null is returned.
+     * 
+     * This class is generally thread safe, but this method is not synchronized with
+     * concurrent write access to avoid performance issues for multithreaded scenarios.
+     * The trade-off for that choice is that in case of concurrent write access, it is
+     * unspecified whether this method returns according to the old state or the new state.
+     * This method is thread safe in the sense that a concurrent write access never exposes
+     * an invalid intermediate state to this read access.
      */
     public static Object getAttr(PyObject ob, byte attr_type) {
-        synchronized (ob) {
-            if (ob.attributes instanceof JyAttribute) {
-                JyAttribute att = (JyAttribute) ob.attributes;
-                while (att != null && att.attr_type < attr_type) {
-                    att = att.getNext();
-                }
-                return att != null && att.attr_type == attr_type ? att.getValue() : null;
+        /*
+        Here used to be a synchronized(ob) block encapsulating the whole method code.
+        It has been removed due to performance issues reported in
+        github.com/jython/jython/issues/360.
+        By making ob.attributes in PyObject and next and value in AttributeLink and
+        TransientAttributeLink volatile, the read access in this method does not
+        fatally interfere with a concurrent write operation.
+        (Write operations continue to be synchronized against each other.)
+        https://docs.oracle.com/javase/specs/jls/se17/html/jls-17.html#jls-17.4.4 states:
+        "A write to a volatile variable v (§8.3.1.4) synchronizes-with all subsequent reads
+         of v by any thread"
+        This induces an x happens-before y relation hb(x, y) according to
+        https://docs.oracle.com/javase/specs/jls/se17/html/jls-17.html#jls-17.4.5:
+        "If an action x synchronizes-with a following action y, then we also have hb(x, y)."
+        Also see https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html,
+        which confirms the anticipated behavior of volatile fields for Java 5 and onwards.
+        Write accesses in the linked attribute list are designed to update the list tail
+        such that a concurrent read access would iterate either through the old state or
+        the new state. No invalid intermediate state is exposed to the read access.
```

Comment:
It will be important that the list node is fully initialised and stable before it can be referenced by the unsynchronised reader in code like this: https://github.com/jython/jython/blob/4c271d88e6a85847b5f2d378cdbc7b3553aab2ae/src/org/python/core/JyAttribute.java#L296-L300
I see that the lines of code here are sequenced so that the publication of the reference to the new node comes last. Now we have to be sure the compiler won't rearrange assignment before construction. ISTR either the JLS or Pugh explicitly mentions this possibility. We have guarantees about writes to an individual variable, but what about actions nearby in the programme order of the thread? I think the generous application of the `volatile` keyword may have achieved that, but as I've said, I find this stuff difficult to reason about securely.

Reply:
Good point about the array. And given that initialisation might skip round the write to `ob.attributes` I'm not convinced it helps. The idea was to reduce the number of volatile fields so we had only one to reason about instead of 3n+1 things.

That's the sort of test I meant, but it might not have to run so long if you could guarantee the threads compete. Inspecting the output of a particular compiler is not good enough. Our correctness has to be based on good theory and a reasonably harsh practical exam.